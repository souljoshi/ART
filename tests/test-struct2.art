/* Test struct behavior in functions */

struct intpair {
    int x;
    int y;
}


/* function that accepts struct by value */
// This shouldn't change the struct
void do_nothing (struct intpair pair)
{
    pair.x = 123;
    pair.y = 42;
}

/* function that accepts struct by reference */
// This should change the struct
void modify (struct intpair& pair)
{
    pair.x = 21;
    pair.y = 32;
}

/* funciton that returns a struct */
struct intpair pair(int x, int y)
{
    struct intpair out;
    out.x = x;
    out.y = y;
    return out;
}

int main() {

    struct intpair a,b,c;

    a.x = b.x = c.x = 0;
    a.y = b.y = c.y = 0;


    printi(a.x);
    printi(a.y);
    printc('\n');
    printi(b.x);
    printi(b.y);
    printc('\n');
    printi(c.x);
    printi(c.y);
    printc('\n');

    do_nothing(a);
    modify(b);
    c = pair(1, -1);

    printi(a.x);
    printi(a.y);
    printc('\n');
    printi(b.x);
    printi(b.y);
    printc('\n');
    printi(c.x);
    printi(c.y);   
}