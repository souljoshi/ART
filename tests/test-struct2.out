0
0

0
0

0
0

0
0

21
32

1
-1